### ISSUE TITLE
Avoiding blocking other threads when waiting for touch

### ISSUE THREAD

**Issue by ronf** [danni/python-pkcs11#86]
Hello,

I'm trying to use python-pkcs11 from a tool which uses asyncio. It's working, but I'm running into a problem when making calls which block, such as when performing a sign operation which requires the user to touch the security token before the sign completes.

To avoid blocking other asyncio coroutines, I'm using asyncio's "run_in_executor" with a thread pool executor. When I test this with other blocking calls such a time.sleep(), the code works properly, running the executor in parallel with operations in other threads, including the main thread's asyncio event loop. However. when I make a PKCS#11 call, even running in an executor seems to block all threads while waiting for the user to touch the token.

I'm thinking maybe that the code which is blocking to wait for the touch operation might be holding onto the global interpreter lock. Is there any way to release that lock while waiting for the user to touch the token and then re-acquire it, so that other threads can make progress while waiting for the user to act?

Short of moving all PKCS#11 operations into their own separate process, do you see any way to prevent user touch operations from blocking all threads?

**Comment by danni** [comment]
So this is going to require a bit of debugging/tracing on your end. It depends greatly on what your PKCS#11 library is doing, e.g. it could be calling join on all threads. `strace` etc will help here.

There's nothing specific I can think of in python-pkcs11 that would be causing it to block. It works in other threaded situations I've used it in (e.g. Django).

Currently python-pkcs11 doesn't do anything to fiddle with the threading options because that was the most universally compatible way to avoid inadvertently sharing sessions between threads or other things that are poorly described in the spec. You could try setting the threading flags in `C_Initialize` (you'll need to build from source) and see if that improves things for you.

**Comment by ronf** [comment]
Thanks for the response. To provide a little more detail, the main program is all running in a single thread, inside an asyncio main loop. To allow blocking operations, the calls to PKCS11 are run in a ThreadPoolExecutor (documented at https://docs.python.org/3/library/concurrent.futures.html). I've used this successfully in the past for code that does things like blocking network I/O, and also tested it with things like sleep(), and it works fine. However, in this case, blocking the thread seems to block the entire process. There's a comment at https://python-pkcs11.readthedocs.io/en/latest/concurrency.html which suggests that this library might be holding the Python global interpreter lock while performing PKCS#11 operations, and that would probably explain what I'm seeing. Unfortunately, this blocks other threads (and all progress in the asyncio event loop), defeating the point of putting the PKCS11 calls into an executor.

This wouldn't be so bad if the blocking was only for long enough to perform a cryptographic operation. However, it also blocks all other threads while waiting for the user to touch their security token, when that feature is enabled. This is a real problem.

Here's some minimal example code that demonstrates the problem:

```python
import asyncio
import pkcs11
import time

from pkcs11 import Attribute, ObjectClass

# Replace this with the path to your PKCS#11 proprietor
lib = pkcs11.lib('/opt/local/lib/libykcs11.dylib')

async def ping():
    while True:
        print('Ping')
        await asyncio.sleep(1)

def sign(key):
    time.sleep(3)
    
    print('Signing...')
    key.sign(b'data')
    print('Done signing')

    time.sleep(3)
    
async def main():
    loop = asyncio.get_running_loop()
    
    ping_task = loop.create_task(ping())

    token = lib.get_token()
    session = token.open(user_pin='123456') # Replace with your PIN
    keys = list(session.get_objects({Attribute.CLASS: ObjectClass.PRIVATE_KEY}))

    while True:
        await loop.run_in_executor(None, sign, keys[0])

asyncio.run(main())
```

The program should be printing "Ping" once a second even while in the middle of calling sign(). However instead, the call to sign() blocks the whole process, and the "Ping" messages only start coming out again when the sign() completes. Note that this does not happen with the calls to time.sleep() before and after sign(), even though they too are blocking calls.

As you can see, this doesn't requite any kind of concurrent calls to pkcs11 to trigger this, or any sharing of token or session objects. I am careful in the real code to only open a session once for any given token to avoid getting errors like the fact that the user has already logged into a token. However, I'm hitting this blocking issue even before I attempt to perform mutiple operations in parallel.

Unfortunately, I have no experience with cython yet, so I haven't been able to entirely make sense of the python-pkcs11 wrapper code. I did find https://speice.io/2019/12/release-the-gil.html, which seems promising. However, so far I haven't identified exactly where PKCS11 would be waiting for the touch operation, which is the main place I'd want to see something like this added.

Any advice would be welcome!

**Comment by danni** [comment]
Hi, so the purpose of that page is not to say it holds the GIL, but that due to the GIL we can avoid most of the thread concurrency problems we'd expect there. However, I admit it could still be holding the lock unintentionally. The problem with PKCS#11 is you're handing quite a lot of the control over to a relatively poorly specified C library, so the first question would be what HSM are you using? I don't have a HSM that requires you to touch to authenticate, so I can't reproduce your issue.

It might be that being more explicit around whether or not we're holding the GIL inside the Cython code is helpful. See https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#acquiring-and-releasing-the-gil -- you'd be best to start in `_pkcs11.pyx` in `Token.open` around the call to `C_Login`, which is what I'm guessing blocks.

Otherwise, it's still worth tracing what your PKCS#11 library is doing. That's pretty easy if it's open source, else you'll need to rely on strace and friends. It's possible that it's deadlocking itself due to the threading mode that it's in.

I'd be keen to see what comes of this, asyncio is a natural interface to the PKCS#11 spec.

**Comment by ronf** [comment]
I'm testing with a Yubikey 5 NFC, which has interfaces for both FIDO2/U2F and PIV (among other things). I had previously implemented an asyncio interface to the FIDO parts of the key using the "fido2" Python module, and that's working great. The executor worked fine there for allowing me to call blocking code from within the asyncio event loop, including things that required device touch.

It's not actually the C_Login operation that's blocking (at least not for long). That doesn't require touch (just the PIN). I think it may be the call to C_SignInit that's actually blocking waiting for the touch. This is called from within the SignMixin class.

Is it possible to add "nogil" to the functions defined in _pkcs11_defn.pxd? In the past, I've made calls into C code like this using ctypes, and that releases the GIL automatically before calling anything marked a CFUNCTYPE for the duration of the call (while things marked as PYFUNCTYPE will hold onto the GIL).

I'll try to get Cython set up here so I can do some experimentation with this. So far, I've just been using the precompiled version of python-pkcs11 from PyPI.

**Comment by danni** [comment]
There is no precompiled version. So Cython should already be in your environment to compile it locally. If you `pip install -e` an editable version of the code directly from git it should work and let you edit things.

I don't think there's a way to define certain calls as always dropping the GIL. Cython avoids auto-dropping the GIL because of the performance hit to drop and require the GIL for what could be a short call.

I notice there's nothing in `C_SignInit` about it being a blocking call (unlike other calls such as when the pin is expected on an external interface). But that's so classic for PKCS#11.

**Comment by ronf** [comment]
I experimented a bit last night on this, and learned a few things:

1) The actual blocking call is in C_Sign, or C_SignFinal when you use the update/final calls. Thinking about it, this makes sense, as the SignInit is just setting up the key & mechanism and the SignUpdate is just updating the hash. The cryptographic operation doesn't happen until you encrypt the final digest, which happens at the end of C_Sign or in the call to C_SignFinal.

2) Since I'm not that familiar with Cython yet, I decided to try and experiment where I replace the sign() call in my test program with a ctypes-based call to C_SignInit() and C_Sign(), leveraging the python-pkcs11 module's code to do all the setup. I'm happy to report that this version of the code does _not_ block all threads! So, this confirms the problem is not anything happening in the underlying PKCS11 C library I'm using, and is instead probably related to Cython not releasing the GIL before making the blocking call. If we can refactor it to do all the operations on the Python variables while holding the GIL but then release the call before calling C_Sign/C_SignFinal, I think that would fix the problem I'm seeing.

There are probably other places in the Cython wrapper code that would need similar treatment (like calls to C_Encrypt, for instance), to really have a full solution here. Basically, anything which involves the library blocking to do user interaction might need a change of this sort.

**Comment by ronf** [comment]
Good news - I've actually managed to get this to work within the existing Cython code! Here's an initial diff that does this for SignMixin. It may be worth using the new C variables I defined in more than just the "nogil" block, but to keep the diff as simple as possible for now, here's what I changed:

```patch
diff --git a/pkcs11/_pkcs11.pyx b/pkcs11/_pkcs11.pyx
index 837696a..399e310 100644
--- a/pkcs11/_pkcs11.pyx
+++ b/pkcs11/_pkcs11.pyx
@@ -927,8 +927,12 @@ class SignMixin(types.SignMixin):
             self.key_type, DEFAULT_SIGN_MECHANISMS,
             mechanism, mechanism_param)
 
+        cdef CK_ULONG session_handle
+        cdef CK_BYTE *data_buffer
+        cdef CK_ULONG data_length
         cdef CK_BYTE [:] signature
         cdef CK_ULONG length
+        cdef CK_RV rv
 
         with self.session._operation_lock:
             assertRV(_funclist.C_SignInit(self.session._handle, mech.data, self._handle))
@@ -938,11 +942,17 @@ class SignMixin(types.SignMixin):
                             data, <CK_ULONG> len(data),
                             NULL, &length))
 
+            session_handle = self.session._handle
+            data_buffer = data
+            data_length = len(data)
             signature = CK_BYTE_buffer(length)
 
-            assertRV(_funclist.C_Sign(self.session._handle,
-                            data, <CK_ULONG> len(data),
-                            &signature[0], &length))
+            with nogil:
+                rv = _funclist.C_Sign(session_handle,
+                                      &data_buffer[0], data_length,
+                                      &signature[0], &length)
+
+            assertRV(rv)
 
             return bytes(signature[:length])
 
@@ -953,8 +963,10 @@ class SignMixin(types.SignMixin):
             self.key_type, DEFAULT_SIGN_MECHANISMS,
             mechanism, mechanism_param)
 
+        cdef CK_ULONG session_handle
         cdef CK_BYTE [:] signature
         cdef CK_ULONG length
+        cdef CK_RV rv
 
         with self.session._operation_lock:
             assertRV(_funclist.C_SignInit(self.session._handle, mech.data, self._handle))
@@ -971,10 +983,14 @@ class SignMixin(types.SignMixin):
             assertRV(_funclist.C_SignFinal(self.session._handle,
                                  NULL, &length))
 
+            session_handle = self.session._handle
             signature = CK_BYTE_buffer(length)
 
-            assertRV(_funclist.C_SignFinal(self.session._handle,
-                                 &signature[0], &length))
+            with nogil:
+                rv = _funclist.C_SignFinal(session_handle,
+                                           &signature[0], &length)
+
+            assertRV(rv)
 
             return bytes(signature[:length])
 
diff --git a/pkcs11/_pkcs11_defn.pxd b/pkcs11/_pkcs11_defn.pxd
index dee6ee7..9169ad7 100644
--- a/pkcs11/_pkcs11_defn.pxd
+++ b/pkcs11/_pkcs11_defn.pxd
@@ -432,7 +432,7 @@ cdef extern from '../extern/cryptoki.h':
                      CK_BYTE *text,
                      CK_ULONG text_len,
                      CK_BYTE *signature,
-                     CK_ULONG *sig_len)
+                     CK_ULONG *sig_len) nogil
 
         CK_RV C_SignUpdate(CK_SESSION_HANDLE session,
                            CK_BYTE *part,
@@ -440,7 +440,7 @@ cdef extern from '../extern/cryptoki.h':
 
         CK_RV C_SignFinal(CK_SESSION_HANDLE session,
                           CK_BYTE *signature,
-                          CK_ULONG *sig_len)
+                          CK_ULONG *sig_len) nogil
 
         CK_RV C_SignRecoverInit(CK_SESSION_HANDLE session,
                                 CK_MECHANISM *mechanism,
```

The test program I pasted above now continues to print "Ping" messages even while sitting in the "Sign" operation waiting for the user to touch the token.

Since the whole block is still protected by the self.session._operation_lock, I think it'll still be able to prevent multiple sign operations from running in parallel and interfering with one another. However, with the GIL no longer held, other non-PKCS11 Python code is free to run in other threads while waiting for this blocking call to complete.

**Comment by danni** [comment]
This is really exciting. Thanks for this. Can you open a PR for this so we can run the changes against the test suite?

`C_Decrypt` also requires access to the secret/private key to make data less protected so would be another candidate for blocking. However, given it's possible that all encryption operations are potentially slow, we should probably make all C calls happen with the GIL released, so we don't block the GIL for any of them while the HSM performs the operation.

It's worth noting that Cython has a PR to add a feature to cause a function to always be called without the GIL (see https://github.com/cython/cython/issues/3443) which could make this work a lot simpler.

**Comment by ronf** [comment]
Sure - would you like me to submit just the Sign case for now in the PR for you to test, or do something wider that covers other encryption related calls? As you said, it might even make sense to cover all calls to the C_* functions, or at least all those which have the potential to be expensive, even though some of them might not block for user interaction.

The cython change you mentioned does sound promising for simplifying this work, if they provide a keyword you can use on a cdef of a function call that automatically releases & reacquires the lock around that call. That said, there might be some cases where we could do better than that using an explicit "with nogil", if there are multiple calls being made in a wrapper like C_SignInit and C_Sign being called multiple times to get the buffer size before doing the actual operation.

Regarding the test suite, is there documentation about running that? I tried running "setup.py test" but I got a bunch of errors, even after setting what I think are all the right environment variables. Mostly it was complaining about operations that my Yubikey didn't support.

**Comment by danni** [comment]
Submit what you've got and let CI run it.

The tests work the best against SoftHSM2. It would be nice to make them run against the YubiKey, but it depends on the capabilities of the HSM. If you can't create temporary keys most of the test suite won't run and will make a mess of your device creating hundreds of keys, and there won't be much point. There are a bunch of test decorators to skip tests that definitely don't work on a given HSM, either through `Mechanisms` or flags, or just by masking libraries that just don't work correctly.

**Comment by ronf** [comment]
Ok - I've submitted pull request #87 with a slightly improved version of the changes I posted above. I was able to mark the assertRV() function as "nogil", so I could check for the return value directly inside the "with" block, and I also used the newly declared variables in some of the prior C_* calls.

I thought about trying to move C_SignInit and the first C_Sign call into the nogil block, but the allocation of the signature byte buffer seems to require the GIL. So, I left it only the final C_Sign/C_SignFinal in the block.

**Comment by danni** [comment]
Thanks for this. I wonder if it's worth adding an asyncio version of the API to the package.

**Comment by ronf** [comment]
Thanks very much for your help, and for the quick turnaround!

Regarding making an async API to this, that's a cool idea but I think the implementation would end up pretty much doing what I'm doing now to run the code in a ThreadPoolExecutor, since the underlying PCKS#11 provider function is still going to block. You could hide this detail by having an async function that returned an asyncio Future which got set when the blocking call finished, but you'd still have to launch a separate thread to run that blocking call.

Asyncio actually makes it pretty easy to use an executor for this. Instead of:

```python
    result = await some_async_func(args)
```

...you just call:

```python
    result = loop.run_in_executor(None, some_blocking_func, *args)
```

...where "loop" is your asyncio event loop.

**Comment by ronf** [comment]
Thanks again for your help on this, Danni!

One last request: Would you be able to include this change in an official release? I'll be checking in the changes  which take advantage of this into the "develop" branch of AsyncSSH shortly, and I'd like to be able to point at a released version of python-pkcs11 which includes these changes in my setup.py before making it a part of a public AsyncSSH release.

**Comment by danni** [comment]
No worries. This is tagged off as v0.7.0.

**Comment by ronf** [comment]
Terrific - thanks very much!

**Comment by ronf** [comment]
The support for this is now available in [AsyncSSH 2.4.0](https://github.com/ronf/asyncssh/releases/tag/v2.4.0) - thanks again!
